File: typescript/leet_code/2597.ts
/*
 *  Copyright ©2010-2024 <a href="https://github.com/jerodg/">JerodG</a>
 *
 *  This program is free software: you can redistribute it and/or modify it under the terms of the
 *  Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
 *  or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 *  even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
 *  for more details.
 *
 *  The above copyright notice and this permission notice shall be included in all copies or
 *  substantial portions of the Software. You should have received a copy of the SSPL along with this
 *  program. If not, see SSPL.
 *
 */

/**
 * Package leet_code
 *
 * This file contains the implementation of the `beautifulSubsets` function.
 * The function calculates the total number of beautiful subsets in a given array of numbers.
 * A subset is considered beautiful if the difference between the maximum and minimum numbers in the
 * subset is equal to `k`.
 */

/**
 * Function to calculate the total number of beautiful subsets in a given array of numbers.
 *
 * @param {number[]} nums - The array of numbers.
 * @param {number} k - The difference between the maximum and minimum numbers in a subset to be considered beautiful.
 * @returns {number} - The total number of beautiful subsets in the array.
 *
 * @example
 *
 * beautifulSubsets([1, 2, 3, 4], 1)
 * // returns 15
 *
 * beautifulSubsets([1, 2, 3, 4], 2)
 * // returns 9
 */
function beautifulSubsets(nums: number[], k: number): number {
    // Initialize total count of beautiful subsets
    let totCount = 1;

    // Initialize a frequency map to store the frequency of each remainder when a number is divided by `k`
    const freqMap: Map<number, Map<number, number>> = new Map();

    // Populate the frequency map
    for (let num of nums) {
        let remainder = num % k;
        if (!freqMap.has(remainder)) {
            freqMap.set(remainder, new Map());
        }
        const remainderMap = freqMap.get(remainder)!;
        remainderMap.set(num, (remainderMap.get(num) || 0) + 1);
    }

    // Calculate the total count of beautiful subsets
    for (let fr of freqMap.values()) {
        const subsets = Array.from(fr.entries()).sort((a, b) => a[0] - b[0]);
        const n = subsets.length;
        let currCount = 1;
        let next1 = 1;
        let next2 = 0;

        for (let i = n - 1; i >= 0; i--) {
            let skip = next1;

            let take = (1 << subsets[i][1]) - 1;

            if (i + 1 < n && subsets[i + 1][0] - subsets[i][0] === k) {
                take *= next2;
            } else {
                take *= next1;
            }

            currCount = skip + take;
            next2 = next1;
            next1 = currCount;
        }

        totCount *= currCount;
    }

    // Subtract 1 from the total count to exclude the empty subset and return the result
    return totCount - 1;
}

File: cpp/leet_code/2597.cpp
/**
* Copyright ©2010-2024 <a href="https://github.com/jerodg/">JerodG</a>
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the
 * Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
 * for more details.
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software. You should have received a copy of the SSPL along with this
 * program. If not, see SSPL.
 */

/**
 * @file leet_code/2597.cpp
 * @brief This file contains the Solution class which is used to solve the problem of finding beautiful subsets.
 *
 * Package leet_code
 */

#include <map>
#include <vector>

/**
 * @class Solution
 * @brief This class is used to solve the problem of finding beautiful subsets.
 *
 * A beautiful subset is defined as a subset of an array where the difference between any two elements is equal to a given difference.
 */
class Solution {
private:
    /**
     * @brief This method is used to count the number of beautiful subsets in a given array.
     *
     * @param subsets A vector of pairs, where each pair represents a subset.
     * @param num_subsets The total number of subsets.
     * @param difference The difference between any two elements in a beautiful subset.
     * @param i The current index in the subsets vector.
     * @param counts A vector to store the count of beautiful subsets for each index.
     * @return The count of beautiful subsets.
     */
    static int count_beautiful_subsets(std::vector<std::pair<int, int>>& subsets, const int num_subsets, const int difference,
        const int i, std::vector<int>& counts) {
        if (i == num_subsets) {
            return 1;
        }

        if (counts[i] != -1) {
            return counts[i];
        }

        // Count the beautiful subsets when the current subset is skipped
        const int skip = count_beautiful_subsets(subsets, num_subsets, difference, i + 1, counts);

        // Count the beautiful subsets when the current subset is taken
        int take = (1 << subsets[i].second) - 1;

        // If the next subset is a beautiful subset with the current subset, multiply the count
        if (i + 1 < num_subsets && subsets[i + 1].first - subsets[i].first == difference) {
            take *= count_beautiful_subsets(subsets, num_subsets, difference, i + 2, counts);
        } else {
            take *= count_beautiful_subsets(subsets, num_subsets, difference, i + 1, counts);
        }

        // Store the count of beautiful subsets for the current index
        return counts[i] = skip + take;
    }

public:
    /**
     * @brief This method is used to find the total number of beautiful subsets in a given array.
     *
     * @param nums The given array.
     * @param difference The difference between any two elements in a beautiful subset.
     * @return The total number of beautiful subsets.
     */
    static int beautifulSubsets(std::vector<int>& nums, const int difference) {
        int tot_count = 1;
        // A map to store the frequency of each number modulo the difference
        std::map<int, std::map<int, int>> freq_map;

        // Populate the frequency map
        for (int& num: nums) {
            freq_map[num % difference][num]++;
        }

        // For each unique number modulo the difference, count the beautiful subsets
        for (auto& fr: freq_map) {
            std::vector<std::pair<int, int>> subsets(fr.second.begin(), fr.second.end());
            std::vector<int> counts(subsets.size(), -1);
            tot_count *= count_beautiful_subsets(subsets, subsets.size(), difference, 0, counts);
        }
        return tot_count - 1;
    }
};
